source/validate_types.cpp

```diff
@@ -0,0 +1,204 @@
+      SpvOpTypeMatrix,
+      SpvOpTypeImage,
+      SpvOpTypeSampler,
+      SpvOpTypeSampledImage,
+      SpvOpTypeArray,
+      SpvOpTypeRuntimeArray,
+      SpvOpTypeStruct,
+      SpvOpTypeOpaque,
+      SpvOpTypePointer,
+      SpvOpTypeFunction,
+      SpvOpTypeEvent,
+      SpvOpTypeDeviceEvent,
+      SpvOpTypeReserveId,
+      SpvOpTypeQueue,
+      SpvOpTypePipe,
+      SpvOpTypeForwardPointer,
+      SpvOpConstantTrue,
+      SpvOpConstantFalse,
+      SpvOpConstant,
+      SpvOpConstantComposite,
+      SpvOpConstantSampler,
+      SpvOpConstantNull,
+      SpvOpSpecConstantTrue,
+      SpvOpSpecConstantFalse,
+      SpvOpSpecConstant,
+      SpvOpSpecConstantComposite,
+      SpvOpSpecConstantOp,
+      SpvOpVariable,
+      SpvOpLine
+    }
+  };
+
+  // clang-format on
+  return moduleOrder;
+}
+}
+
+namespace libspirv {
+
+ValidationState_t::ValidationState_t(spv_diagnostic* diag, uint32_t options)
+    : diagnostic_(diag),
+      instruction_counter_(0),
+      defined_ids_{},
+      unresolved_forward_ids_{},
+      validation_flags_(options),
+      operand_names_{},
+      module_layout_order_stage_(0),
+      current_layout_stage_(ModuleLayoutSection::kModule) {}
+
+spv_result_t ValidationState_t::defineId(uint32_t id) {
+  if (defined_ids_.find(id) == end(defined_ids_)) {
+    defined_ids_.insert(id);
+  } else {
+    return diag(SPV_ERROR_INVALID_ID) << "ID cannot be assigned multiple times";
+  }
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidationState_t::forwardDeclareId(uint32_t id) {
+  unresolved_forward_ids_.insert(id);
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidationState_t::removeIfForwardDeclared(uint32_t id) {
+  unresolved_forward_ids_.erase(id);
+  return SPV_SUCCESS;
+}
+
+void ValidationState_t::assignNameToId(uint32_t id, string name) {
+  operand_names_[id] = name;
+}
+
+string ValidationState_t::getIdName(uint32_t id) const {
+  string out = to_string(id);
+  if (operand_names_.find(id) != end(operand_names_)) {
+    out += "[" + operand_names_.at(id) + "]";
+  }
+  return out;
+}
+
+size_t ValidationState_t::unresolvedForwardIdCount() const {
+  return unresolved_forward_ids_.size();
+}
+
+vector<uint32_t> ValidationState_t::unresolvedForwardIds() const {
+  vector<uint32_t> out(begin(unresolved_forward_ids_),
+                       end(unresolved_forward_ids_));
+  return out;
+}
+
+bool ValidationState_t::isDefinedId(uint32_t id) const {
+  return defined_ids_.find(id) != end(defined_ids_);
+}
+
+bool ValidationState_t::is_enabled(spv_validate_options_t flag) const {
+  return (flag & validation_flags_) == flag;
+}
+
+// Increments the instruction count. Used for diagnostic
+int ValidationState_t::incrementInstructionCount() {
+  return instruction_counter_++;
+}
+
+ModuleLayoutSection ValidationState_t::getLayoutStage() const {
+  return current_layout_stage_;
+}
+
+void ValidationState_t::progressToNextLayoutStageOrder() {
+  if (++module_layout_order_stage_ >= GetModuleOrder().size()) {
```

Comment:
Quibble.   module_layout_order_stage_ is beyond size, continuing to call this function will continue to increment the value.  Eventually it will wrap around and get back to the start.
Better to increment the value only when it's less than the limit.  Yes, it will make slightly more code, but will be slightly more safe.